agent/stats/engine.go

```diff
@@ -108,31 +108,51 @@ func (engine *DockerStatsEngine) MustInit(taskEngine ecsengine.TaskEngine, clust
 	return engine.Init()
 }
 
-// Init initializes the docker client's event engine. This must be called
-// to subscribe to the docker's event stream.
+// Init subscribes to the container change event stream.
 func (engine *DockerStatsEngine) Init() error {
-	engine.ctx, engine.unsubscribeContainerEvents = context.WithCancel(context.TODO())
-	err := engine.openEventStream()
+	// Subscribe to the container change event stream
+	err := engine.containerChangeEventStream.Subscribe(containerChangeHandler, engine.handleDockerEvents)
 	if err != nil {
-		return err
+		return fmt.Errorf("Failed to subscribe to container change event stream, err %v", err)
 	}
 
 	go engine.listContainersAndStartEventHandler()
+	go func() {
+		// Waiting for the event stream to close
+		ctx := engine.containerChangeEventStream.Context()
+		c, _ := context.WithCancel(ctx)
```

Comment:
It creates a child context, which will be closed when the parent's context's Done is closed or call the cancel, here if the ctx(from the event stream) is closed, the child c->Done() will also closed, we don't need the cancel to close the child context.